File: src/accessiweather/display/presentation/source_attribution.py
"""Source attribution presentation helpers."""

from __future__ import annotations

from ...models import WeatherData
from .models import SourceAttributionPresentation

_WET_CONDITION_MARKERS = (
    "thunderstorm",
    "t-storm",
    "rain",
    "shower",
    "drizzle",
    "snow",
    "sleet",
    "hail",
    "freezing rain",
)
_DRY_CONDITION_MARKERS = (
    "clear",
    "sunny",
    "cloudy",
    "overcast",
    "fair",
)


def build_source_attribution(weather_data: WeatherData) -> SourceAttributionPresentation | None:
    """Build source attribution presentation for transparency."""
    attribution = weather_data.source_attribution
    if not attribution:
        return None

    source_names = {
        "nws": "National Weather Service",
        "openmeteo": "Open-Meteo",
        "visualcrossing": "Visual Crossing",
        "pirateweather": "Pirate Weather",
    }

    contributing = [
        source_names.get(source, source.title())
        for source in sorted(attribution.contributing_sources)
    ]
    failed = [
        source_names.get(source, source.title()) for source in sorted(attribution.failed_sources)
    ]
    incomplete = list(weather_data.incomplete_sections)

    if contributing:
        summary = f"Data from: {', '.join(contributing)}"
        if failed:
            summary += f". Unavailable: {', '.join(failed)}"
    elif failed:
        summary = f"Sources unavailable: {', '.join(failed)}"
    else:
        summary = ""

    disagreement_note = _build_source_disagreement_note(weather_data, source_names)
    if disagreement_note:
        summary = f"{summary}. {disagreement_note}" if summary else disagreement_note

    aria_parts = []
    if contributing:
        aria_parts.append(f"Weather data provided by {', '.join(contributing)}")
    if failed:
        aria_parts.append(f"Data unavailable from {', '.join(failed)}")
    if incomplete:
        aria_parts.append(f"Missing sections: {', '.join(incomplete)}")
    if disagreement_note:
        aria_parts.append(disagreement_note)
    aria_label = ". ".join(aria_parts) if aria_parts else "Weather data source information"

    return SourceAttributionPresentation(
        contributing_sources=contributing,
        failed_sources=failed,
        incomplete_sections=incomplete,
        summary_text=summary,
        aria_label=aria_label,
    )


def _build_source_disagreement_note(
    weather_data: WeatherData, source_names: dict[str, str]
) -> str | None:
    """Return a short note when obvious condition sources disagree."""
    attribution = weather_data.source_attribution
    if attribution is None or weather_data.current is None:
        return None

    current_text = _clean_condition_text(weather_data.current.condition)
    current_state = _precipitation_state(current_text)
    if current_text is None or current_state is None:
        return None

    hourly_period = (
        weather_data.hourly_forecast.periods[0]
        if weather_data.hourly_forecast and weather_data.hourly_forecast.periods
        else None
    )
    hourly_text = _clean_condition_text(
        getattr(hourly_period, "short_forecast", None) if hourly_period else None
    )
    hourly_state = _precipitation_state(hourly_text)

    minutely = weather_data.minutely_precipitation
    minutely_text = _clean_condition_text(getattr(minutely, "summary", None))
    minutely_state = _minutely_precipitation_state(minutely)

    parts: list[str] = []
    field_sources = attribution.field_sources
    current_source = _format_source_name(field_sources.get("condition"), source_names)
    hourly_source = _format_source_name(field_sources.get("hourly_source"), source_names)
    minutely_source = _format_source_name(
        field_sources.get("minutely_precipitation") or field_sources.get("hourly_summary"),
        source_names,
    )

    if hourly_text and hourly_state is not None and hourly_state != current_state:
        parts.append(f"hourly forecast from {hourly_source} says {hourly_text}")

    if minutely_text and minutely_state is not None and minutely_state != current_state:
        parts.append(
            f"minute-by-minute precipitation outlook from {minutely_source} says {minutely_text}"
        )

    if not parts:
        return None

    return (
        f"Data note: current conditions from {current_source} report {current_text}; "
        f"{'; '.join(parts)}."
    )


def _clean_condition_text(value: str | None) -> str | None:
    """Normalize condition text for display while preserving user-facing wording."""
    if not isinstance(value, str):
        return None
    cleaned = value.strip()
    return cleaned.rstrip(".") if cleaned else None


def _precipitation_state(value: str | None) -> str | None:
    """Classify obvious wet/dry condition wording; return None when ambiguous."""
    if value is None:
        return None
    normalized = value.casefold()
    if any(marker in normalized for marker in _WET_CONDITION_MARKERS):
        return "wet"
    if any(marker in normalized for marker in _DRY_CONDITION_MARKERS):
        return "dry"
    return None


def _minutely_precipitation_state(minutely: object | None) -> str | None:
    """Classify minutely precipitation data from explicit points or summary text."""
    if minutely is None:
        return None

    points = getattr(minutely, "points", None)
    if points:
        for point in points:
            intensity = getattr(point, "precipitation_intensity", None)
            probability = getattr(point, "precipitation_probability", None)
            if (isinstance(intensity, int | float) and intensity > 0) or (
                isinstance(probability, int | float) and probability > 0
            ):
                return "wet"
        return "dry"

    return _precipitation_state(_clean_condition_text(getattr(minutely, "summary", None)))


def _format_source_name(source: str | None, source_names: dict[str, str]) -> str:
    """Return a readable source label, falling back to a generic label."""
    if not source:
        return "another source"
    return source_names.get(source, source.title())

File: rust/crates/aw-core/src/display/source_attribution.rs
//! Data source attribution.
//!
//! Port of `display/presentation/source_attribution.py`.

use crate::display::models::SourceAttributionPresentation;
use crate::display::pyfmt::title;
use crate::model::{MinutelyPrecipitationForecast, WeatherData};

const WET_MARKERS: [&str; 9] = [
    "thunderstorm",
    "t-storm",
    "rain",
    "shower",
    "drizzle",
    "snow",
    "sleet",
    "hail",
    "freezing rain",
];
const DRY_MARKERS: [&str; 5] = ["clear", "sunny", "cloudy", "overcast", "fair"];

fn source_name(source: &str) -> String {
    match source {
        "nws" => "National Weather Service".into(),
        "openmeteo" => "Open-Meteo".into(),
        "pirateweather" => "Pirate Weather".into(),
        other => title(other),
    }
}

/// `build_source_attribution`.
///
/// Python lists `incomplete_sections` in set iteration order, which varies
/// between runs; here they are sorted.
pub fn build_source_attribution(data: &WeatherData) -> Option<SourceAttributionPresentation> {
    let attribution = data.source_attribution.as_ref()?;
    let contributing: Vec<String> = attribution
        .contributing_sources
        .iter()
        .map(|s| source_name(s))
        .collect();
    let failed: Vec<String> = attribution
        .failed_sources
        .iter()
        .map(|s| source_name(s))
        .collect();
    let incomplete: Vec<String> = data.incomplete_sections.iter().cloned().collect();

    let mut summary = if !contributing.is_empty() {
        let mut s = format!("Data from: {}", contributing.join(", "));
        if !failed.is_empty() {
            s.push_str(&format!(". Unavailable: {}", failed.join(", ")));
        }
        s
    } else if !failed.is_empty() {
        format!("Sources unavailable: {}", failed.join(", "))
    } else {
        String::new()
    };

    let note = disagreement_note(data);
    if let Some(n) = &note {
        summary = if summary.is_empty() {
            n.clone()
        } else {
            format!("{summary}. {n}")
        };
    }

    let mut aria = Vec::new();
    if !contributing.is_empty() {
        aria.push(format!("Weather data provided by {}", contributing.join(", ")));
    }
    if !failed.is_empty() {
        aria.push(format!("Data unavailable from {}", failed.join(", ")));
    }
    if !incomplete.is_empty() {
        aria.push(format!("Missing sections: {}", incomplete.join(", ")));
    }
    if let Some(n) = note {
        aria.push(n);
    }
    let aria_label = if aria.is_empty() {
        "Weather data source information".into()
    } else {
        aria.join(". ")
    };

    Some(SourceAttributionPresentation {
        contributing_sources: contributing,
        failed_sources: failed,
        incomplete_sections: incomplete,
        summary_text: summary,
        aria_label,
    })
}

/// `_clean_condition_text`.
fn clean(value: Option<&str>) -> Option<String> {
    let cleaned = value?.trim();
    (!cleaned.is_empty()).then(|| cleaned.trim_end_matches('.').to_string())
}

/// `_precipitation_state`: "wet", "dry" or ambiguous.
fn precipitation_state(value: Option<&str>) -> Option<&'static str> {
    let v = value?.to_lowercase();
    if WET_MARKERS.iter().any(|m| v.contains(m)) {
        Some("wet")
    } else if DRY_MARKERS.iter().any(|m| v.contains(m)) {
        Some("dry")
    } else {
        None
    }
}

fn minutely_state(minutely: Option<&MinutelyPrecipitationForecast>) -> Option<&'static str> {
    let m = minutely?;
    if !m.points.is_empty() {
        let wet = m.points.iter().any(|p| {
            p.precipitation_intensity.is_some_and(|i| i > 0.0)
                || p.precipitation_probability.is_some_and(|p| p > 0.0)
        });
        return Some(if wet { "wet" } else { "dry" });
    }
    precipitation_state(clean(m.summary.as_deref()).as_deref())
}

fn format_source(source: Option<&String>) -> String {
    match source.filter(|s| !s.is_empty()) {
        Some(s) => source_name(s),
        None => "another source".into(),
    }
}

/// `_build_source_disagreement_note`.
fn disagreement_note(data: &WeatherData) -> Option<String> {
    let attribution = data.source_attribution.as_ref()?;
    let current = data.current.as_ref()?;
    let current_text = clean(current.condition.as_deref())?;
    let current_state = precipitation_state(Some(&current_text))?;

    let hourly_text = data
        .hourly_forecast
        .as_ref()
        .and_then(|h| h.periods.first())
        .and_then(|p| clean(p.short_forecast.as_deref()));
    let hourly_state = precipitation_state(hourly_text.as_deref());
    let minutely = data.minutely_precipitation.as_ref();
    let minutely_text = minutely.and_then(|m| clean(m.summary.as_deref()));
    let minutely_state = minutely_state(minutely);

    let fs = &attribution.field_sources;
    let current_source = format_source(fs.get("condition"));
    let hourly_source = format_source(fs.get("hourly_source"));
    let minutely_source = format_source(
        fs.get("minutely_precipitation")
            .filter(|s| !s.is_empty())
            .or(fs.get("hourly_summary")),
    );

    let mut parts = Vec::new();
    if let (Some(text), Some(state)) = (&hourly_text, hourly_state) {
        if state != current_state {
            parts.push(format!("hourly forecast from {hourly_source} says {text}"));
        }
    }
    if let (Some(text), Some(state)) = (&minutely_text, minutely_state) {
        if state != current_state {
            parts.push(format!(
                "minute-by-minute precipitation outlook from {minutely_source} says {text}"
            ));
        }
    }
    if parts.is_empty() {
        return None;
    }
    Some(format!(
        "Data note: current conditions from {current_source} report {current_text}; {}.",
        parts.join("; ")
    ))
}
